### PR TITLE
JIT: infer ranges from small int type operations

### LINKED ISSUE #21841: The JIT should elide bounds checks when indexing into a byte array with length = byte.MaxValue.
I am seeing an extra bounds check (see full disassembly below) which is forcing me to use unsafe code:
```text
00007ffc`f32c0669 4181f800010000  cmp     r8d,100h
00007ffc`f32c0670 7324            jae     00007ffc`f32c0696
```

A `byte` cannot be >= 256 and hence the check is unnecessary.

**Benchmark:**
```C#

private class ConfigWithCustomEnvVars : ManualConfig
{
    private const string JitNoInline = "COMPlus_TieredCompilation";

    public ConfigWithCustomEnvVars()
    {
        Add(Job.Core
            .With(new[] { new EnvironmentVariable(JitNoInline, "0") })
            .WithWarmupCount(3)
            .WithTargetCount(5)
            .With(CsProjCoreToolchain.From(NetCoreAppSettings.NetCoreApp30)));
    }
}

private byte[] _tempArray;

[GlobalSetup]
public void Setup()
{
    _tempArray = new byte[100];
    for (int i = 0; i < 100; i++)
        _tempArray[i] = (byte)'a';
}

private static ReadOnlySpan<byte> AllowList => new byte[256] {
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1,
    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1,
    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1,
    0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
};

[MethodImpl(MethodImplOptions.AggressiveInlining)]
private static bool NeedsEscaping_new(byte value)
    => Unsafe.Add(ref MemoryMarshal.GetReference(AllowList), value) == 0;

[MethodImpl(MethodImplOptions.AggressiveInlining)]
private static bool NeedsEscaping(byte value) => AllowList[value] == 0;

[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static int NeedsEscaping(ReadOnlySpan<byte> value)
{
    int idx;
    for (idx = 0; idx < value.Length; idx++)
    {
        if (NeedsEscaping(value[idx]))
        {
            goto Return;
        }
    }

    idx = -1; // all characters allowed

Return:
    return idx;
}

[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static int NeedsEscaping_new(ReadOnlySpan<byte> value)
{
    int idx;
    for (idx = 0; idx < value.Length; idx++)
    {
        if (NeedsEscaping_new(value[idx]))
        {
            goto Return;
        }
    }

    idx = -1; // all characters allowed

Return:
    return idx;
}


[Benchmark(Baseline = true)]
public int TestingBoundsCheck()
{
    return NeedsEscaping(_tempArray);
}

[Benchmark]
public int TestingBoundsCheck_new()
{
    return NeedsEscaping_new(_tempArray);
}
```

``` ini

BenchmarkDotNet=v0.11.3, OS=Windows 10.0.17763.195 (1809/October2018Update/Redstone5)
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-preview-010006
  [Host] : .NET Core 3.0.0-preview-27218-01 (CoreCLR 4.6.27217.02, CoreFX 4.7.18.61304), 64bit RyuJIT
  Core   : .NET Core 3.0.0-preview-27218-01 (CoreCLR 4.6.27217.02, CoreFX 4.7.18.61304), 64bit RyuJIT

Job=Core  EnvironmentVariables=COMPlus_TieredCompilation=0  Runtime=Core  
Toolchain=.NET Core 3.0  IterationCount=5  WarmupCount=3  

```
|                 Method | Formatted | SkipValidation |     Mean |    Error |    StdDev | Ratio |
|----------------------- |---------- |--------------- |---------:|---------:|----------:|------:|
|     TestingBoundsCheck |     False |           True | 81.91 ns | 1.515 ns | 0.3935 ns |  1.00 |
| TestingBoundsCheck_new |     False |           True | 74.59 ns | 2.463 ns | 0.6397 ns |  0.91 |

```text

System.Text.JsonLab.Benchmarks.JsonWriterPerf
TestingBoundsCheck .NET Core 3.0.0-preview-27218-01 (CoreCLR 4.6.27217.02, CoreFX 4.7.18.61304), 64bit RyuJIT	TestingBoundsCheck_new .NET Core 3.0.0-preview-27218-01 (CoreCLR 4.6.27217.02, CoreFX 4.7.18.61304), 64bit RyuJIT

00007ffc`f32c0640 System.Text.JsonLab.Benchmarks.JsonWriterPerf.TestingBoundsCheck()
00007ffc`f32c0640 4883ec28        sub     rsp,28h
00007ffc`f32c0644 90              nop
00007ffc`f32c0645 488b4140        mov     rax,qword ptr [rcx+40h]
00007ffc`f32c0649 4885c0          test    rax,rax
00007ffc`f32c064c 7506            jne     00007ffc`f32c0654
00007ffc`f32c064e 33d2            xor     edx,edx
00007ffc`f32c0650 33c9            xor     ecx,ecx
00007ffc`f32c0652 eb07            jmp     00007ffc`f32c065b
00007ffc`f32c0654 488d5010        lea     rdx,[rax+10h]
00007ffc`f32c0658 8b4808          mov     ecx,dword ptr [rax+8]
00007ffc`f32c065b 33c0            xor     eax,eax
00007ffc`f32c065d 85c9            test    ecx,ecx
00007ffc`f32c065f 7e2b            jle     00007ffc`f32c068c
00007ffc`f32c0661 4c63c0          movsxd  r8,eax
00007ffc`f32c0664 460fb60402      movzx   r8d,byte ptr [rdx+r8]
00007ffc`f32c0669 4181f800010000  cmp     r8d,100h
00007ffc`f32c0670 7324            jae     00007ffc`f32c0696
00007ffc`f32c0672 4d63c0          movsxd  r8,r8d
00007ffc`f32c0675 49b9d4e3012d7c010000 mov r9,offset Benchmarks!COM+_Entry_Point  (Benchmarks+0x1ce3d4) (0000017c`2d01e3d4)
00007ffc`f32c067f 43803c0800      cmp     byte ptr [r8+r9],0
00007ffc`f32c0684 740b            je      00007ffc`f32c0691
00007ffc`f32c0686 ffc0            inc     eax
00007ffc`f32c0688 3bc1            cmp     eax,ecx
00007ffc`f32c068a 7cd5            jl      00007ffc`f32c0661
00007ffc`f32c068c b8ffffffff      mov     eax,0FFFFFFFFh
00007ffc`f32c0691 4883c428        add     rsp,28h
00007ffc`f32c0695 c3              ret
00007ffc`f32c0696 e895dac85f      call    coreclr!GetCLRRuntimeHost+0x881a0 (00007ffd`52f4e130)
00007ffc`f32c069b cc              int     3
00007ffc`f32c069c 190401          sbb     dword ptr [rcx+rax],eax
00007ffc`f32c069f 000442          add     byte ptr [rdx+rax*2],al
00007ffc`f32c06a2 0000            add     byte ptr [rax],al
00007ffc`f32c06a4 400000          add     byte ptr [rax],al
00007ffc`f32c06a7 00d8            add     al,bl
00007ffc`f32c06a9 ac              lods    byte ptr [rsi]
00007ffc`f32c06aa 40              ???
00007ffc`f32c06ab f3fc            rep cld
00007ffc`f32c06ad 7f00            jg      00007ffc`f32c06af
00007ffc`f32c06af 000f            add     byte ptr [rdi],cl
00007ffc`f32c06b1 1f              ???
00007ffc`f32c06b2 440000          add     byte ptr [rax],r8b
00007ffc`f32c06b5 c3              ret

No ILOffsetMap found
System.Text.JsonLab.Benchmarks.JsonWriterPerf.NeedsEscaping(System.ReadOnlySpan`1)
System.Text.JsonLab.Benchmarks.JsonWriterPerf.NeedsEscaping(Byte)
System.Text.JsonLab.Benchmarks.JsonWriterPerf.get_AllowList()


00007ffc`f32d0640 System.Text.JsonLab.Benchmarks.JsonWriterPerf.TestingBoundsCheck_new()
00007ffc`f32d0640 0f1f440000      nop     dword ptr [rax+rax]
00007ffc`f32d0645 488b4140        mov     rax,qword ptr [rcx+40h]
00007ffc`f32d0649 4885c0          test    rax,rax
00007ffc`f32d064c 7506            jne     00007ffc`f32d0654
00007ffc`f32d064e 33d2            xor     edx,edx
00007ffc`f32d0650 33c9            xor     ecx,ecx
00007ffc`f32d0652 eb07            jmp     00007ffc`f32d065b
00007ffc`f32d0654 488d5010        lea     rdx,[rax+10h]
00007ffc`f32d0658 8b4808          mov     ecx,dword ptr [rax+8]
00007ffc`f32d065b 33c0            xor     eax,eax
00007ffc`f32d065d 85c9            test    ecx,ecx
00007ffc`f32d065f 7e22            jle     00007ffc`f32d0683
00007ffc`f32d0661 4c63c0          movsxd  r8,eax
00007ffc`f32d0664 460fb60402      movzx   r8d,byte ptr [rdx+r8]
00007ffc`f32d0669 4d63c0          movsxd  r8,r8d
00007ffc`f32d066c 49b9d4e380fa5c010000 mov r9,offset Benchmarks!COM+_Entry_Point  (Benchmarks+0x1ce3d4) (0000015c`fa80e3d4)
00007ffc`f32d0676 43803c0800      cmp     byte ptr [r8+r9],0
00007ffc`f32d067b 740b            je      00007ffc`f32d0688
00007ffc`f32d067d ffc0            inc     eax
00007ffc`f32d067f 3bc1            cmp     eax,ecx
00007ffc`f32d0681 7cde            jl      00007ffc`f32d0661
00007ffc`f32d0683 b8ffffffff      mov     eax,0FFFFFFFFh
00007ffc`f32d0688 c3              ret
00007ffc`f32d0689 0000            add     byte ptr [rax],al
00007ffc`f32d068b 0019            add     byte ptr [rcx],bl
00007ffc`f32d068d 0000            add     byte ptr [rax],al
00007ffc`f32d068f 004000          add     byte ptr [rax],al
00007ffc`f32d0692 0000            add     byte ptr [rax],al
00007ffc`f32d0694 0000            add     byte ptr [rax],al
00007ffc`f32d0696 0000            add     byte ptr [rax],al
00007ffc`f32d0698 80da41          sbb     dl,41h
00007ffc`f32d069b f3fc            rep cld
00007ffc`f32d069d 7f00            jg      00007ffc`f32d069f
00007ffc`f32d069f 000f            add     byte ptr [rdi],cl
00007ffc`f32d06a1 1f              ???
00007ffc`f32d06a2 440000          add     byte ptr [rax],r8b
00007ffc`f32d06a5 c3              ret
00007ffc`f32d06a6 0000            add     byte ptr [rax],al

No ILOffsetMap found
System.Text.JsonLab.Benchmarks.JsonWriterPerf.NeedsEscaping_new(System.ReadOnlySpan`1)
System.Text.JsonLab.Benchmarks.JsonWriterPerf.NeedsEscaping_new(Byte)
System.Text.JsonLab.Benchmarks.JsonWriterPerf.get_AllowList()
System.Runtime.CompilerServices.Unsafe.Add(!!0 ByRef, Int32)

```

cc @AndyAyersMS, @GrabYourPitchforks, @dotnet/jit-contrib

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Have range check use the ranges for small int types if it has no better
information.

Fixes #21841

### CODE REVIEW

**Comment by mikedn** [comment]
I suppose you could use `INT8_MIN`, `INT8_MAX` and so on but it's probably a bit of an abuse as those are relevant to C/C++ types and not JIT's IR types.

Or you could "steal" the code from GenIntCastDesc that computes min/max from the type size and signedness: https://github.com/dotnet/coreclr/blob/ca65764c029f2dac6f4a187694f4232eec9b1115/src/jit/codegenlinear.cpp#L1989-L1991 but I don't think it's worth the trouble.

**Comment by AndyAyersMS** [comment]
Test failures seem unrelated, will rerun a few...

@dotnet-bot retest OSX10.12 x64 Checked Innerloop Build and Test
@dotnet-bot retest Windows_NT arm Cross Checked Innerloop Build and Test

**Comment by AndyAyersMS** [comment]
OSX now showing "typical" failures in PAL file tests:
```
Test Result (5 failures / +5)
file_io.GetFileSize.test1
file_io.GetFileSizeEx.test1
file_io.SetFilePointer.test5
file_io.SetFilePointer.test6
file_io.SetFilePointer.test7
```
will give it one more try....

@dotnet-bot retest OSX10.12 x64 Checked Innerloop Build and Test

[Ubuntu arm](https://ci.dot.net/job/dotnet_coreclr/job/master/job/arm_cross_checked_ubuntu_innerloop_tst_prtest/7339/) failure in "dynamo" GC test (linking to #17129)
```
14:22:16         /ssd/j/workspace/dotnet_coreclr/master/arm_cross_checked_ubuntu_innerloop_tst_prtest/bin/tests/Linux.arm.Checked/Tests/Core_Root/corerun dynamo.exe 1000 40 191919
14:22:16          
14:22:16         Total amount of RandomNode Memory: 
14:22:16         12487500
14:22:16          
14:22:16         Running Finalize Test...
14:22:16          
14:22:16         Deleting Node:
14:22:16         0
14:22:16         Deleting Node:
14:22:16         19
14:22:16         The registered number: 50
14:22:16         The analyzed number: 42
14:22:16         The registered number and analyzed number are not equal
14:22:16         
14:22:16         Test Failed with seed: 191919
```

@dotnet-bot retest Ubuntu arm Cross Checked Innerloop Build and Test

**Comment by erozenfeld** [comment]
We have MAX_SHORT_AS_INT and MIN_SHORT_AS_INT consts in jit.h (currently unused). Perhaps we can add constants for ushort, byte, and ubyte there.

**Comment by erozenfeld** [comment]
LGTM

**Comment by GrabYourPitchforks** [comment]
@AndyAyersMS, in your code gen you have the following.

```asm
       0FB6140A             movzx    rdx, byte  ptr [rdx+rcx]
       4863D2               movsxd   rdx, edx
```

Is it possible to get the JIT to drop the `movsxd` instruction? It seems redundant.

**Comment by AndyAyersMS** [comment]
The jit disassembler is not quite right here -- actual disassembly is
```asm
       0FB6140A             movzx    edx, byte  ptr [rdx+rcx]    // dest is edx, not rdx
       4863D2               movsxd   rdx, edx
```
(encoding would need to be `48 0fb6140a` to be `movzx rdx ...`)

And while it seems like it ought to be simple to fix this, there are various issues...

Early on there are two uses of the load result -- one for the bounds check and one for the array element address, of type `int` and `long` respectively. The stack type is `int` so that's how we initially model the result of the load. One could argue `int` is a good choice here as the bounds check will happen at least as often as the address computation.

Later during optimization we eliminate the bounds check (which this PR enables) and only the `long` use in the element address computation remains, so we could in principle try and change the type to match the remaining use (that is, push the "cast to long" onto the load), but there's nothing in the jit to do that. The def and use trees are split and there is some cruft in between. And we don't try to restitch broken trees.

We don't catch this later on after lowering even though def and use trees end up adjacent because the jit does not model the implicit upper half zeroing on x64 for 32 bit register destinations (that is, we don't realize the `movzx edx` we emit does appropriate extension and the cast is not needed) and the machine-level IR we producing is not all that amenable to peephole opts that might let us catch this sort of thing.

**Comment by pentp** [comment]
If the JIT did model/track the implicit upper half zeroing on x64, it would allow most sign-extensions to be eliminated from array/span accesses also. I tried to get something working in #21553, but at the moment it looks a bit too complicated for my limited JIT development skill. While replacing `movsx` with `mov` works, it doesn't gain much (the temp registers are still allocated).

File: src/jit/rangecheck.cpp [hunk 1]
```diff
@@ -1135,6 +1143,29 @@ Range RangeCheck::ComputeRange(BasicBlock* block, GenTree* expr, bool monotonic
...
     }
+    else if (varTypeIsSmallInt(expr->TypeGet()))
+    {
+        switch (expr->TypeGet())
+        {
+            case TYP_UBYTE:
+                range = Range(Limit(Limit::keConstant, 0), Limit(Limit::keConstant, 255));
+                break;
+            case TYP_BYTE:
+                range = Range(Limit(Limit::keConstant, -127), Limit(Limit::keConstant, 128));
```
[the remark this answers is not in the document]
Comment (AndyAyersMS) [reply src/jit/rangecheck.cpp]: Yeah, it should be... thanks for spotting it.

**Comment by erozenfeld** [comment]
We have GetLowerBoundForIntegralType and GetUpperBoundForIntegralType that could've been used here.

**Pull request merged**